### DISCUSSION TITLE
I have an error .

### DISCUSSION THREAD

**Discussion by AdamMusa** [serverpod/serverpod#1037] [Q&A]
I used Serverpod in my project and I  uploaded correctly my  files(image or pdf, docs etc) but when I want to get the public url Iw have  this error in my terminal with my public url  printed from server side

http://localhost:8080/serverpod_cloud_storage?method=file&path=%2FVolumes%2FMacintosh+HD%2FUsers%2Fmacbookpro%2FDownloads%2FImage%2Flogin.png
METHOD CALL: uploaderFile.getUrl duration: 5ms numQueries: 0 authenticatedUser: null
2023-06-24 14:17:54.458969Z Internal server error. Zoned exception.
Converting object to an encodable object failed: Instance of '_Uri'
#0      _JsonStringifier.writeObject (dart:convert/json.dart:793:7)
#1      _JsonStringStringifier.printOn (dart:convert/json.dart:982:17)
#2      _JsonStringStringifier.stringify (dart:convert/json.dart:967:5)
#3      JsonEncoder.convert (dart:convert/json.dart:345:30)
#4      JsonCodec.encode (dart:convert/json.dart:232:37)
#5      jsonEncode (dart:convert/json.dart:114:10)
#6      SerializationManager.encode (package:serverpod_serialization/src/serialization.dart:146:12)
#7      Server._handleRequest (package:serverpod/src/server/server.dart:292:53)
<asynchronous suspension>

And I cannot get this public url from client side.  Can someone help me please

**Comment by dannejosefsson** [comment]
You have probably solved this by now, but I ended up in the same situation following the[Uploading files documentation](https://docs.serverpod.dev/concepts/file-uploads).

I ended up changing my endpoint method to return `Future<String?>` instead of `Future<Uri?>`. On the frontend side, you can use `Uri.parse(publicUrlString)` or `Uri.tryParse(publicUrlString)` to get the actual object if you want to have easy access to the `Uri` parts.

The underlying problem is that `Uri` interface doesn't have the `fromJson` and `toJson` methods needed for serialisation. You can implement it yourself by following [Custom Serialization]https://docs.serverpod.dev/concepts/serialization). The result would look something like:

```
import 'package:serverpod_serialization/serverpod_serialization.dart';

class ServerpodUri implements ProtocolSerialization {
  Uri uri;

  ServerpodUri._(this.uri);

  factory ServerpodUri.fromUri(Uri uri) = ServerpodUri._;

  @override
  toJsonForProtocol() => toJson();

  factory ServerpodUri.fromJson(String uri) {
    return ServerpodUri.fromUri(Uri.parse(uri));
  }

  String toJson() {
    return uri.toString();
  }
}
```

It had been neat if it would be possible to register `JsonConverter`s like [freezed does it](https://pub.dev/packages/freezed#fromjsontojson) instead of implementing wrapper classes like this. Any takes on that @SandPod?

**Reply by SandPod** [reply]
That would have been really neat!

This stems from us playing catch-up to cover all dart types in our JsonConverted code: https://github.com/serverpod/serverpod/blob/bee2cfe19bac000f3f09517d4a66b439f83e8cc3/packages/serverpod_serialization/lib/src/serialization.dart#L198-L227

To allow this we would need to make these methods non static and then allow developers to register their own `JsonConverter` additions when instantiating the `Protocol()` class.
